### PR TITLE
chore: prevent Copilot CLI 5 MB request overflows

### LINKED ISSUE #1995: Copilot CLI: prevent 5 MB CAPI request overflows
## Problem

Long Copilot CLI maintenance and GUI-validation sessions repeatedly fail with:

> The request is too large to send through CAPI Responses. (5.2 MB request; 5 MB limit)

Local transcript analysis found two concrete contributors:

- One failed session retained eight PNG binary assets totaling 2,732,303 decoded bytes (about 3.48 MiB after base64/event encoding), leaving too little room for normal system, tool, and conversation content.
- Another failed autonomous session accumulated thousands of turns and tens of MiB of tool/model event data between only two compactions.

The machine also defaults to `long_context`, while Copilot CLI compaction is token-triggered. The fixed 5 MB serialized-request ceiling can therefore be reached before token utilization triggers compaction.

Upstream `github/copilot-cli#3767` was marked fixed in v1.0.62, but new July 2026 reports reproduce compaction failures on v1.0.71 image-heavy sessions. Configurable earlier compaction remains open in `github/copilot-cli#1688`.

## Acceptance criteria

- Repository sessions default to the normal context tier unless explicitly overridden.
- Tool-driven image inspection has a conservative cumulative per-agent byte budget and fails open if the guard cannot run.
- Large diffs, logs, and screenshots remain in isolated subagent contexts instead of being copied repeatedly into the coordinator history.
- Daily maintenance avoids unbounded main-session ingestion while still completing the full routine.
- Automated tests cover the context guard and run in CI.
- Contributor documentation includes prevention and recovery steps for the 5 MB failure.

## Non-goals

- Modifying Copilot CLI internals or bypassing the server-side 5 MB limit.
- Rewriting or deleting existing session history.
- Preventing direct user attachment of multiple images, which current hooks cannot inspect before attachment interning.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

### PR DESCRIPTION
**Pull request by laqieer**
## Summary

- default trusted FEBuilderGBA Copilot sessions to the normal context tier so token compaction starts earlier than the fixed 5 MB transport ceiling
- add a fail-open `preToolUse` image-budget guard with validated Bash/PowerShell wrappers, per-session private state, and crash-safe OS kernel locking
- keep screenshots, full diffs, logs, and untrusted-PR safety screens inside fresh bounded child contexts instead of the coordinator history
- document prevention, unattended prompt-mode activation, explicit reset, and recovery for oversized CAPI requests

## Plan Reference

Implements the accepted Plan v3.1 from https://github.com/laqieer/FEBuilderGBA/issues/1995#issuecomment-5016952653

Plan Review Gate approval: https://github.com/laqieer/FEBuilderGBA/issues/1995#issuecomment-5017118199

## Scope

Closes #1995

Supersedes #1996 with a fresh replacement branch. The prior published history was not rewritten; follow-up commits fix its commitlint failure, remove unsafe automatic state pruning, and address every review finding.

The guard intentionally covers image files read through Copilot CLI's `view` tool. Direct human attachments and unrelated screenshot tools remain upstream CLI concerns; workflow isolation and the repository `contextTier` default cover those paths.

## Test plan

- [x] `python -m unittest discover -s scripts/tests -p "test_copilot_context_guard.py" -v` — 87 tests passed on Windows; 10 platform-specific cases skipped locally and run in the existing Ubuntu/macOS/Windows CI matrix
- [x] validated fail-open wrapper behavior for missing interpreter, missing/unreadable guard, child exit 1, malformed/empty exit-2 output, and valid deny exit 2
- [x] validated cross-process lock contention, normal release, and crash release using `msvcrt` on Windows; `fcntl` runs in Unix CI
- [x] validated that arbitrarily old fallback state is never pruned automatically and only explicit deletion or `/new` resets the budget
- [x] parsed `.github/hooks/copilot-context-budget.json` and `.github/copilot/settings.json` as JSON and `.github/workflows/crossplatform.yml` as YAML
- [x] trusted-checkout live dispatch with `GITHUB_COPILOT_PROMPT_MODE_REPO_HOOKS=true`: default budget allowed a real `view` call; a 1-byte budget denied the same image through `.github_hooks_copilot-context-budget.json`
- [x] ran `git diff --check` and confirmed a clean branch working tree and a commit subject below 100 characters

## Known limitations

- Repository `model`, `effortLevel`, and `contextTier` overrides apply only in a trusted working directory.
- Repository hooks in external `copilot -p` launches require `GITHUB_COPILOT_PROMPT_MODE_REPO_HOOKS=true`; structural child-context rules remain the fallback when hooks are unavailable.
- Guard state is deliberately conservative across compaction and is not automatically pruned; start `/new` or explicitly remove that session's guard state to reset the image budget.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

### CODE REVIEW

**Comment by laqieer** [comment]
## Cross-Model Review Board — PR #1997

**Verdict: APPROVE (3/3, zero Blocking findings)**

Reviewed exact head: `1296c17874f7a706397b32ef607273f4bbeea763`.

### Consensus

- The fail-open hook and wrappers correctly distinguish infrastructure failures from a validated deny decision.
- `fcntl`/`msvcrt` kernel locking handles contention and crash release without stale-lock deletion races.
- Private per-session/per-user state is atomic, never automatically pruned, and resets only through `/new` or explicit deletion.
- The normal repository context tier, prompt-mode hook opt-in, bounded child reports, visual-review isolation, and untrusted-PR safety-screen rules match the accepted Plan v3.1.
- The single commit has a compliant subject; every test-plan item is checked and specific.
- This is a `chore` with no GUI files, so screenshots and a GUI Test Report are not required.

### Non-blocking follow-up

- Complete the final required CI matrix before merge.
- Perform a one-time interactive `/env` plus allow/deny hook smoke before relying on enforcement; structural isolation remains the safe fallback if repository hooks are unavailable.

Review Board: gpt-5.5, claude-opus-4.8, gemini-3.5-flash
Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/copilot-context-guard.sh [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+# or the fail-closed preToolUse contract would deny unrelated view() calls.
+if [ "$status" -ne 0 ]; then
+  printf '{}'
+  exit 0
+fi
+
+if [ -z "$output" ]; then
+  printf '{}'
+else
+  printf '%s' "$output"
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/copilot-context-guard.sh]: Fixed in a0c160c69. The Bash wrapper now normalizes every child exit-0 result to the fixed `{}` abstention and never forwards child stdout. Regression tests cover malformed, partial, arbitrary, and forged-deny exit-0 output.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/copilot-context-guard.ps1 [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+if ($status -ne 0) {
+    Write-Output '{}'
+    exit 0
+}
+
+if ([string]::IsNullOrEmpty($joined)) {
+    Write-Output '{}'
+} else {
+    Write-Output $joined
+}
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/copilot-context-guard.ps1]: Fixed in a0c160c69. The PowerShell wrapper now always emits `{}` on child exit 0 instead of forwarding output. Matching regression tests cover malformed, partial, arbitrary, and forged-deny output.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/copilot_context_guard.py [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+        st = os.stat(abs_path)
+    except OSError:
+        return FALL_THROUGH, 0
+
+    size = st.st_size
+    mtime_ns = getattr(st, "st_mtime_ns", int(st.st_mtime * 1e9))
+    fingerprint = _normalized_fingerprint(abs_path, size, mtime_ns)
+
+    session_id = payload.get("sessionId")
+    state_dir, is_fallback = _resolve_state_dir(session_id)
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/copilot_context_guard.py:496]: Fixed in a0c160c69. `run()` now abstains before any state resolution when `sessionId` is missing, empty, or non-string. Tests verify invalid IDs create and charge no state.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: README.md [hunk 4]
```diff
@@ -803,6 +803,55 @@ See
...
+  `GITHUB_COPILOT_PROMPT_MODE_REPO_HOOKS=true` is set (the literal string `"true"`, not `1`).
+- `.github/copilot/settings.json` ships `{"contextTier": "default"}`. Per official Copilot CLI docs,
+  a repository-level `contextTier` setting takes precedence over a user-level setting **only when
+  the working directory is trusted** (untrusted directories fall back to the user-level setting).
+  Repo scope generally outranks user scope in the settings merge order (`user` → `repo` → `local`),
+  so shipping this file makes `default` context tier apply automatically for any trusted checkout of
+  this repo, without requiring each contributor to set it manually. If you deliberately want
+  `long_context` for your own session despite this file, override it at session/user scope (e.g. the
+  `--context long_context` CLI flag, or `/settings set contextTier long_context` in an interactive
+  session) — session/local overrides still win over the repo default.
```
[the remark this answers is not in the document]
Comment (laqieer) [reply README.md]: Fixed in a0c160c69. README now documents the valid higher-precedence local override `/settings --local contextTier long_context` and the one-session launch flag `copilot --context long_context`; it no longer suggests the invalid user-scope command.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/copilot_context_guard.py [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+    to walk up through -- and ``chmod 0700`` -- the shared parent directory
+    itself, which the first user to run the guard would then lock every
+    other local UID out of. Because this root's own name is already
+    user-specific, it is always safe to privatize in full without ever
+    touching anything actually shared between users (``tempfile.gettempdir()``
+    itself is never a target).
+    """
+    return os.path.join(
+        tempfile.gettempdir(), "{0}-{1}".format(FALLBACK_SUBDIR_NAME, _current_user_token())
+    )
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/copilot_context_guard.py]: Fixed in a0c160c69 and hardened further in 60f9272d8. Fallback state no longer uses shared OS temp; it lives under `$COPILOT_HOME/context-guard/<sessionId>`. The guard rejects symlinks, non-directories, wrong POSIX ownership, all non-FileNotFound lstat errors, and re-verifies type/ownership after creation before chmod/use. Dedicated race and permission-error tests are included.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/tests/test_copilot_context_guard.py [hunk 6]
```diff
@@ -0,0 +1,1328 @@
...
+                self.assertEqual(stat.S_IMODE(os.stat(root).st_mode), 0o700)
+                self.assertEqual(stat.S_IMODE(os.stat(state_dir).st_mode), 0o700)
+            # $COPILOT_HOME (fake_home) itself must never be chmod'ed --
+            # only the context-guard root and session subdir beneath it.
+            if not sys.platform.startswith("win"):
+                home_mode = stat.S_IMODE(os.stat(fake_home).st_mode)
+                self.assertNotEqual(
+                    home_mode, 0o700,
+                    "$COPILOT_HOME itself must never be chmod'ed to 0700 by the guard",
+                )
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/tests/test_copilot_context_guard.py:681]: Fixed in 75caef1e4. The POSIX mode-preservation test now explicitly sets the fake COPILOT_HOME to 0755 before invoking the guard and asserts that exact mode remains unchanged.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/copilot_context_guard.py [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+    session-budget state to an attacker-controlled location, e.g. a
+    symlink planted by another local user at a predictable path. Any
+    failure to even stat the path is also treated as unsafe -- this
+    function fails open by refusing to use an uncertain path, never by
+    trusting one.
+    """
+    try:
+        st = os.lstat(path)
+    except OSError:
+        return True  # does not exist yet: safe to create fresh
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/copilot_context_guard.py]: Already fixed in 60f9272d8. `_is_safe_private_dir_target()` now treats only `FileNotFoundError` as absent; every other lstat error returns false/fails open. Post-creation lstat/type/ownership verification and dedicated PermissionError/symlink-race tests are included.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/tests/test_copilot_context_guard.py [hunk 8]
```diff
@@ -0,0 +1,1328 @@
...
+                python_dirs.add(os.path.dirname(found))
+        original_path = os.environ.get("PATH", "")
+        sep = os.pathsep
+        filtered = sep.join(
+            part for part in original_path.split(sep)
+            if os.path.normcase(os.path.normpath(part)) not in
+            {os.path.normcase(os.path.normpath(d)) for d in python_dirs}
+        )
+        out, code = self._run(_payload(tool_name="bash"), env={"PATH": filtered})
+        self.assertEqual((out.strip(), code), ("{}", 0))
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/tests/test_copilot_context_guard.py]: Fixed in 75caef1e4. Both missing-interpreter tests now use a guaranteed-empty PATH plus a real view payload and assert that no state file is created, proving no interpreter ran. The Bash wrapper no longer depends on external dirname, and PowerShell is launched via its resolved absolute path.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/tests/test_copilot_context_guard.py [hunk 9]
```diff
@@ -0,0 +1,1429 @@
...
+        attacker-controlled directory appears in place of a real directory
+        exactly when ``os.makedirs()`` is called. ``os.makedirs(...,
+        exist_ok=True)`` would otherwise silently accept this (it uses the
+        symlink-following ``os.path.isdir()`` internally to decide
+        "already exists, fine") -- the mandatory post-creation ``os.lstat``
+        re-verification must catch and reject it instead, before any
+        chmod or use.
+        """
+        fake_home = tempfile.mkdtemp(prefix="ctxguard-toctou-home-")
+        attacker_target = tempfile.mkdtemp(prefix="ctxguard-toctou-attacker-")
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/tests/test_copilot_context_guard.py:860]: Fixed in e8074807e. The attacker target now starts at an explicit 0755 baseline and the race test asserts that exact mode remains unchanged.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/copilot_context_guard.py [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+            post_st = os.lstat(directory)
+        except OSError:
+            return False
+        if not _is_owned_real_dir(post_st):
+            return False
+        if not sys.platform.startswith("win"):
+            try:
+                os.chmod(directory, 0o700)
+            except OSError:
+                pass
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/copilot_context_guard.py]: Fixed in e8074807e. On POSIX, any chmod(0700) failure now returns false, so run() abstains and persists no state. Unit and end-to-end forced-chmod-failure tests cover the contract.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/tests/test_copilot_context_guard.py [hunk 11]
```diff
@@ -0,0 +1,1429 @@
...
+                python_dirs.add(os.path.dirname(found))
+        original_path = os.environ.get("PATH", "")
+        sep = os.pathsep
+        filtered = sep.join(
+            part for part in original_path.split(sep)
+            if os.path.normcase(os.path.normpath(part)) not in
+            {os.path.normcase(os.path.normpath(d)) for d in python_dirs}
+        )
+        out, code = self._run(_payload(tool_name="bash"), env={"PATH": filtered})
+        self.assertEqual((out.strip(), code), ("{}", 0))
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/tests/test_copilot_context_guard.py]: Fixed in 75caef1e4. PowerShell is launched by its resolved absolute path while PATH is a guaranteed-empty directory; a real view payload plus absence of the state marker proves no Python interpreter ran.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: README.md [hunk 12]
```diff
@@ -803,6 +803,58 @@ See
...
+- `.github/hooks/copilot-context-budget.json` registers a best-effort `preToolUse` command hook
+  (`scripts/copilot_context_guard.py`, plus `scripts/copilot-context-guard.sh` /
+  `scripts/copilot-context-guard.ps1` platform wrappers) that tracks cumulative bytes of image files
+  read through the `view` tool in a session and denies a further read once a conservative budget
+  (default 1,250,000 bytes, overridable via `COPILOT_CONTEXT_GUARD_BUDGET_BYTES`) would be exceeded.
+  It is fail-open by design: any uncertain, malformed, or infrastructure-failure path abstains
+  (`{}`, exit 0) rather than blocking unrelated work, and only a definitive cumulative overflow denies
+  (JSON `permissionDecision: "deny"` + a reason, exit 2). Repository hooks only execute when the
+  session has repo-hook execution enabled; verify with `/env` that
+  `GITHUB_COPILOT_PROMPT_MODE_REPO_HOOKS=true` is set (the literal string `"true"`, not `1`).
```
[the remark this answers is not in the document]
Comment (laqieer) [reply README.md]: Fixed in e8074807e and completed class-wide in a1360c31b. All four runnable external copilot -p examples and the workflow summary table now establish GITHUB_COPILOT_PROMPT_MODE_REPO_HOOKS=true before launch, with the PowerShell $env: equivalent documented beside them.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/copilot_context_guard.py [hunk 13]
```diff
@@ -0,0 +1,546 @@
...
+            state, was_corrupt = _load_state(state_path)
+            if was_corrupt:
+                # Uncertain state: fail open without persisting anything.
+                return FALL_THROUGH, 0
+
+            entries = state["entries"]
+            if fingerprint in entries:
+                entries[fingerprint]["last_seen"] = time.time()
+                _save_state_atomically(state_dir, state_path, state)
+                return FALL_THROUGH, 0
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/copilot_context_guard.py]: Fixed in 4f6cc7376. Every authorized image read now charges its file size, including repeated identical fingerprints. A denied repeat does not mutate state; fingerprint entries retain only read_count/last_seen metadata.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: scripts/tests/test_copilot_context_guard.py [hunk 14]
```diff
@@ -0,0 +1,1574 @@
...
+        with open(state_path) as fh:
+            state = json.load(fh)
+        self.assertEqual(state["total_bytes"], 900_000)
+        self.assertEqual(len(state["entries"]), 1)
+
+    def test_dedupe_does_not_double_count_identical_read(self):
+        image = self.make_image("same.png", 900_000)
+        guard.run(_payload(tool_args={"path": image}))
+        out2, code2 = guard.run(_payload(tool_args={"path": image}))
+        self.assertEqual((out2, code2), ("{}", 0))
```
[the remark this answers is not in the document]
Comment (laqieer) [reply scripts/tests/test_copilot_context_guard.py]: Fixed in 4f6cc7376. The old dedupe test was replaced: a second 900 KB read under the 1.25 MB budget is denied, while smaller repeated reads increment total_bytes and read_count until the threshold.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: .github/hooks/copilot-context-budget.json [hunk 15]
```diff
@@ -0,0 +1,14 @@
+{
+  "version": 1,
+  "hooks": {
+    "preToolUse": [
+      {
+        "type": "command",
+        "bash": "bash scripts/copilot-context-guard.sh",
+        "powershell": "& \".\\scripts\\copilot-context-guard.ps1\"",
```
[the remark this answers is not in the document]
Comment (laqieer) [reply .github/hooks/copilot-context-budget.json:8]: Verified live from the trusted PR worktree with `GITHUB_COPILOT_PROMPT_MODE_REPO_HOOKS=true`. With the default budget, Copilot CLI executed `view` successfully and replied ALLOW. With `COPILOT_CONTEXT_GUARD_BUDGET_BYTES=1`, the same real `view` call completed as denied by `.github_hooks_copilot-context-budget.json` and replied DENY. The PR test plan now records this runtime evidence.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

File: .github/workflows/crossplatform.yml [hunk 16]
```diff
@@ -41,6 +41,9 @@ jobs:
       working-directory: tools/gba-playtest
       run: python -m pytest -q
 
+    - name: Run copilot-context-guard contract tests (issue #1995)
+      run: python -m unittest discover -s scripts/tests -p "test_copilot_context_guard.py" -v
```
[the remark this answers is not in the document]
Comment (laqieer) [reply .github/workflows/crossplatform.yml:45]: The PR body is already updated for the current head: the exact unittest command reports 87 passing tests with 10 platform-gated skips on Windows, and the trusted live allow/deny dispatch is recorded. The current-head rerun completed successfully after 4f6cc7376.

Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)

**Comment by laqieer** [comment]
## Cross-Model Review Board — Final PR #1997 Review

**Verdict: APPROVE (3/3, zero Blocking findings)**

Reviewed exact final head: `4f6cc7376d14bd38a53adb5c1018d7ac26777dbf`.

### Final evidence

- OpenAI, Anthropic, and Google independently approved the current head.
- Required and advisory CI are green across all 27 checks, including Ubuntu, macOS, and Windows guard tests.
- The current suite records 87 passing tests with 10 platform-gated skips locally; all platform-specific paths passed in CI.
- Trusted-checkout runtime dispatch verified both paths: a normal-budget real `view` call succeeded, while a 1-byte budget was denied by `.github_hooks_copilot-context-budget.json`.
- All 30 inline review comments are resolved; issue-level comments and top-level reviews contain no outstanding action.
- The branch is current with `origin/master`, commitlint passes, every PR test-plan item is checked, and no GUI files changed (`chore` screenshots/GUI report remain optional).

The implementation now charges repeated image reads, normalizes all exit-0 wrapper output, abstains on invalid session IDs, uses crash-safe kernel locks, stores fallback state only in a verified user-owned Copilot cache, never prunes session state automatically, and fails open on every uncertain storage/permission path.

Review Board: gpt-5.5, claude-opus-4.8, gemini-3.5-flash
Copilot CLI: 1.0.72-1
Model: GPT-5.6 Sol (gpt-5.6-sol)
